### ISSUE TITLE
also test against environment modules (Tcl) 4.0.0

### ISSUE THREAD

**Issue by boegel** [easybuilders/easybuild-framework#2341]
https://modules.readthedocs.io/en/stable/NEWS.html

**Comment by ebagrenrut** [comment]
Modules 4 is based on the Tcl Modules implementation (1.923), however its command interfaces have been changed to be compatible with the C implementation that it replaces. As a result, neither setting `EASYBUILD_MODULES_TOOL=EnvironmentModulesC` nor `EASYBUILD_MODULES_TOOL=EnvironmentModulesTcl` works with Modules 4. In the former case, EasyBuild fails to determine the version of the Modules implementation because the `VERSION_REGEXP` in the `EnvironmentModulesC` class does not match the output of 'module --version'. In the latter case, the `COMMAND` `modulecmd.tcl` set in the `EnvironmentModulesTcl` class is not found, as that is not put in the `PATH` by the Modules 4 installation; if one does put `/usr/share/Modules/libexec` into their `PATH`, EasyBuild still fails because the behavior of `modulecmd.tcl` has been changed to match `modulecmd` from the C implementation.

A 'quick' work-around for those using Modules 4 now is to simply fix the `VERSION_REGEXP` in the `EnvironmentModulesC` class, setting it to:

`VERSION_REGEXP = r'^Modules\s+Release\s+(?P<version>\d\S*)\s'`

However that isn't likely a good long-term fix, as Modules 4 is actually a pure Tcl implementation and not a C implementation. Likely an additional class will need to be created to handle Modules 4+.

**Comment by ebagrenrut** [comment]
I've submitted a pull request with a first attempt at supporting Modules 4. #2365

**Comment by boegel** [comment]
Thanks a lot for your efforts on this @ebagrenrut, much appreciated!

**Comment by boegel** [comment]
@ebagrenrut I just installed Modules v4.0.0 to play around with and test your contribution in #2365, but I'm not getting a `modulecmd` at all:

```
$ ls /tmp/$USER/mod4/bin
add.modules envml       mkroot
$ ls /tmp/$USER/mod4/libexec
modulecmd-compat modulecmd.tcl
```

So, you approach in #2365 to leverage `EnvironmentModulesC` isn't going to work, we should leverage `EnvironmentModulesTcl` instead? Probably also other changes are needed, since the way `modulecmd.tcl` has changed...

**Comment by boegel** [comment]
@xdelaruelle Can you shed some light on the above?

**Comment by xdelaruelle** [comment]
@boegel Sure! From what I see I also think it should leverage `EnvironmentModulesTcl` rather `EnvironmentModulesC`. Especially the `set_path_env_var` and `remove_module_path` defs from the Tcl-specific class will also be useful for Modules >=4.

The `modulecmd.tcl` script is installed in `libexec` dir and it is intended to be reached through a shell alias or function. On RH-like systems, a `modulecmd` link onto `modulecmd.tcl` is setup in  `bindir`. This link is made with the `alternatives` tool.

From my understanding of the `ModulesTool` class, the module script is found by searching `COMMAND` in PATHs or by using filename pointed by the environment variable defined in `COMMAND_ENVIRONMENT`. Is there other means ?

Something may be missing in Modules 4.0 currently to be correctly detected by EasyBuild. Hopefully 4.1 will be out in 1 or 2 weeks and I can add the required stuff for a proper detection.

By the way, regarding the new `EnvironmentModules` class name, Modules now follows the semver approach for version numbering and I am pretty sure a 5.0 will be released in 2018. So maybe including the number `4` in the class name may at some point be misleading for the users.

**Comment by ebagrenrut** [comment]
@boegel , @xdelaruelle Thanks for clarifying. I am indeed running on RedHat, thus the /usr/bin/modulecmd link was created. I have no preference as to the name of the new class (if there even needs to be one). Feel free to make whatever changes you see fit to my PR (or deny it and start anew); unfortunately I won't be able to work on it until later in January.

**Comment by ebagrenrut** [comment]
How about these ideas for a new class:

- Class name is `EnvironmentModules`
- `COMMAND` is set to the equivalent of `${MODULESHOME}/libexec/modulecmd.tcl`
- `VERSION_REGEXP` is `r'^Modules\s+Release\s+(?P<version>\d\S*)\s'`
- `REQ_VERSION` set to `4.0.0`

**Comment by ebagrenrut** [comment]
OK, my proposed class works on my RHEL systems, so I've committed it. Unfortunately, Travis seems to be blowing up when trying to install some python modules to test on Python 2.6, which I assume is not related to my code.

**Comment by boegel** [comment]
Thanks a lot for the feedback @xdelaruelle!

The expectation is indeed that `modulecmd.tcl` is available either via `$PATH` or full path via an environment variable. I know neither are currently in place for Modules v4, but the same goes for older versions of environment modules, so people need to include the location of the actual command (`modulecmd` or `modulecmd.tcl`) to `$PATH` to make EasyBuild happy (see also the note in http://easybuild.readthedocs.io/en/latest/Installation.html#required-modules-tool).

The other suggestions are very useful too, thanks again!

**Comment by boegel** [comment]
@ebagrenrut Issue with failing tests should be fixed by #2367. I'll send a pull request to your branch to also enable testing EasyBuild against Modules v4.

**Comment by boegel** [comment]
@ebagrenrut PR to also test EasyBuild against Modules v4 is available at https://github.com/ebagrenrut/easybuild-framework/pull/1, if you merge that your PR #2367 will be updated to include it.

There seems to be an issue with the `module purge` command as called by EasyBuild, leading to an error like this when `modulecmd.tcl python purge` is called:

```
ERROR: Unexpected number of args for 'purge' command
```

Any idea what's going on with this @xdelaruelle?

**Comment by xdelaruelle** [comment]
@boegel `purge` command does not accept any argument and it seems it has detected more than 0 argument in the situation you describe.

Looking at the `purge` def, it seems an empty argument is passed to the `run_module` method:

    self.run_module('purge', '')

Maybe that is what is causing the issue

**Comment by ebagrenrut** [comment]
I can reproduce the behavior using modulscmd.tcl directly:

>% tclsh $MODULESHOME/libexec/modulecmd.tcl python purge ''
raise RuntimeError('modulecmd.tcl: 1 error(s) detected!')
ERROR: Unexpected number of args for 'purge' command

**Comment by boegel** [comment]
@xdelaruelle Nice catch! There's indeed no reason to pass an empty argument to `purge`. It seems like other modules tools just ignore the additional empty argument, while Modules v4 is more picky about it (which actually makes sense to me).

Fixed with https://github.com/ebagrenrut/easybuild-framework/pull/2

**Comment by ebagrenrut** [comment]
@boegel I think there is an assumption in the test_load method of the ModulesTest class that is probably not true for Modules 4. The test loads the GCC/4.6.4 modules, then unsets $EBROOTGCC. A comment indicates that EBROOTGCC will be reset if the GCC/4.6.4 module is loaded again. Modules 4 doesn't reset EBROOTGCC (I've confirmed by simply trying it in my environment), and I think that may be by design.

@xdelaruelle Please correct me if I'm wrong, but if one unsets an environment variable that a module set, and then want that environment variable to be reset to the value specified in the module file, one currently needs to do a module reload, or a module switch. I suspect that removing the 'feature' in which loading a module for a second time would reload the module was a design decision.

@boegel I guess the question is, is this functionality that EasyBuild relies on?

**Comment by xdelaruelle** [comment]
@ebagrenrut If a module is already loaded, new `module load` on this module wil be ignored. So if one variable set by a module is unset, a new module load on this module will not restore it.

This is currently made this way not to mess the path reference counters handled by commands like `append-path`. But I have seen also in other situations that one should be able to reload with a `load` command an already loaded module, especially to restore volatile environment stuff that are lost in sub-shells (like shell aliases). This change is targeted for 4.2 release, which should be out during 2018 spring.

**Comment by boegel** [comment]
@ebagrenrut That particular check was added as a part of a test for the `allow_reload` option of the `load` method implemented by EasyBuild.

By default EasyBuild *allows* reloading of already loaded modules, but in some particular circumstances that needs to be disabled (which makes EasyBuild first check via `module list` before loading a module).

I don't think EasyBuild assumes anywhere that already loaded modules get reloaded, only in this particular test it matters.

So, I think we can just disable this particular check when Modules v4 is being used as modules tool, which is done in https://github.com/ebagrenrut/easybuild-framework/pull/3 .

**Comment by xdelaruelle** [comment]
FYI, starting with version ``4.1`` of Modules, a ``modulecmd`` wrapper script will be installed in binary directory. In addition a ``MODULES_CMD`` environment variable will be set, at initialization time, to the fully qualified path of the module script.

**Comment by boegel** [comment]
@xdelaruelle That's useful, but we'll probably stick to our current approach in #2365, i.e. finding `modulecmd.tcl` via `$MODULESHOME`, since that also works with Modules 4.0.

Hopefully that will stay in place?

**Comment by xdelaruelle** [comment]
Of course, ``$MODULESHOME/libexec/modulecmd.tcl`` will also do the job in future releases.

**Comment by boegel** [comment]
fixed in #2365, thanks for your help @ebagrenrut and @xdelaruelle!
